doc/catalog-format-spec.md

```diff
@@ -547,6 +606,171 @@ available.'case(toolchain)'.user = false
  - `word-size`: architecture word size. Currently supported values are:
    `bits-32`, `bits-64`, `bits-unknown`
 
+## Using crate configuration
+
+`Alire` provides a mechanism for crates to expose a list of variables that can
+be set by other crates depending on them. The configuration variables will then
+be converted to Ada, C and GPR source files that can be used to change the
+behavior or feature set of the code.
+
+Let's start with a simple example. A crate named `test` can print debug log on
+the console. However printing on the console has a performance impact, for an
+embedded project it can even have a significant code size impact. Therefore it
+would be best if this logging can be disabled/enabled at compile time.
+
+To achieve this, a crate maintainer can define a configuration variable in the
+crate manifest `alire.toml`. The definition will be like so:
+```toml
+[config_variables]
+Enable_Logs = {type = "Boolean", default = false}
+```
+A single variable of type `Boolean` with a default value of `false`.
+
+From this definition, `Alire` will generate various source files, including an
+Ada package specification:
+
+```ada
+package Test_Config is
+   Enable_Logs : constant Boolean := False;
+end Test_Config;
+```
+
+In the crate source code, this configuration package can be used like so:
+```ada
+   if Test_Config.Enable_Logs then
+      Print_Log ("This is a log message.");
+   end if;
+```
+
+If one of the crates depending on `test` sets the configuration variable to
+`true`, e.g.:
+
+```toml
+[config_settings]
+test.Enable_Logs = true
+```
+
+The constant value will change in the generated configuration package:
+```ada
+package Test_Config is
+   Enable_Logs : constant Boolean := True;
+end Test_Config;
+```
+Which will enable logging in the `test` crate.
+
+It is possible for multiple depending crates to set `test.Enable_Logs` to the
+same value, however if two depending crates set the variable to a different
+value then the configuration is invalid and `Alire` will print an error. If no
+depending crates set the `test.Enable_Logs` variable, then its default value is
+used.
+
+### When to use crate configuration?
+
+Usually when something has to be static or known at compiler-time, either for
+performance or memory usage.
+
+### When _not_ to use crate configuration?
+
+When the Ada languages provides a better alternative. There are many ways to
+provide an Ada API that will result in compile time optimization or static
+memory usage.
```

Comment:
Should we mention `gprbuild -X` flags here? Simple cases can be done by selecting different versions of a spec file through a scenario variable.

Reply:
We will see how this evolves but crate config should replace scenario variables in most cases.

Reply:
OK.